### PR TITLE
Fixes small evaluation bugs related to casts, types, etc

### PR DESCRIPTION
**Pull request by johnedquinn**
## Description
- Fixes small evaluation bugs related to casts, types, etc
- This is in relation to the removal of v0.x APIs from the v1 branch
- Updated PartiQLResult to return Datum, not PartiQLValue

## License Information

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: partiql-spi/src/main/java/org/partiql/eval/value/Datum.java [hunk 1]
```diff
@@ -596,14 +603,82 @@ static Datum string(@NotNull String value) {
...
+        }
+        if (value.length() < length) {
+            newValue = String.format("%-" + length + "." + length + "s", value);
+        } else if (value.length() == length) {
+            newValue = value;
+        } else {
+            newValue = value.substring(0, length);
+        }
+        return new DatumString(newValue, PType.varchar(length));
+    }
```
Comment (alancai98) [inline partiql-spi/src/main/java/org/partiql/eval/value/Datum.java:667]: Should reference char in some places
```suggestion
    /**
     *
     * @param value the string to place in the char
     * @return a char value with a default length of 255
     */
    @NotNull
    static Datum character(@NotNull String value) {
        return character(value, 255);
    }

    /**
     *
     * @param value the string to place in the char
     * @return a char value
     */
    @NotNull
    static Datum character(@NotNull String value, int length) {
        String newValue;
        if (length <= 0) {
            throw new DataException("CHAR of length " + length + " not allowed.");
        }
        if (value.length() < length) {
            newValue = String.format("%-" + length + "." + length + "s", value);
        } else if (value.length() == length) {
            newValue = value;
        } else {
            newValue = value.substring(0, length);
        }
        return new DatumString(newValue, PType.character(length));
    }
```

**Review by alancai98 — COMMENTED** [review]
May also be helpful to give a bit more description on the bugs or have corresponding tests showing the now-fixed behavior.

File: partiql-spi/src/main/kotlin/org/partiql/value/ion/IonDatum.kt [hunk 2]
```diff
@@ -161,15 +163,35 @@ public class IonDatum private constructor(value: AnyElement, type: PType) : Datu
...
+                    null -> TimeZone.UnknownTimeZone
+                    else -> TimeZone.UtcOffset.of(ts.zHour, ts.zMinute)
+                }
+                DateTimeValue.time(ts.hour, ts.minute, ts.second, tz)
+            }
+            else -> super.getTime()
+        }
     }
 
+    // TODO: Handle struct notation
```
[the remark this answers is not in the document]
Comment (johnedquinn) [reply partiql-spi/src/main/kotlin/org/partiql/value/ion/IonDatum.kt:189]: The `$timestamp::{ year: 2024, month: 12, ... }` notation of timestamps for PartiQL.

File: partiql-eval/src/main/kotlin/org/partiql/eval/internal/operator/rex/CastTable.kt [hunk 3]
```diff
@@ -614,7 +700,7 @@ internal object CastTable {
 
     private fun datumInt(value: BigDecimal): Datum {
         val int = try {
-            value.setScale(0, RoundingMode.DOWN).intValueExact()
+            value.setScale(0, RoundingMode.HALF_EVEN).intValueExact()
```
[the remark this answers is not in the document]
Comment (johnedquinn) [reply partiql-eval/src/main/kotlin/org/partiql/eval/internal/operator/rex/CastTable.kt:703]: Just what I saw when porting the evaluation tests. Here's the evaluating compiler [code](https://github.com/partiql/partiql-lang-kotlin/blob/306ec7c1dc143660d6d861cc9626cbc3ac444ec6/partiql-lang/src/main/kotlin/org/partiql/lang/eval/ExprValueExtensions.kt#L342), and [PostgreSQL behavior](https://www.db-fiddle.com).

Agreed regarding the rounding behavior -- though, we'll get there when we get there.

File: partiql-spi/src/main/java/org/partiql/spi/value/Datum.java [hunk 4]
```diff
@@ -596,14 +603,82 @@ static Datum string(@NotNull String value) {
...
+            throw new DataException("CHAR of length " + length + " not allowed.");
+        }
+        if (value.length() < length) {
+            newValue = String.format("%-" + length + "." + length + "s", value);
+        } else if (value.length() == length) {
+            newValue = value;
+        } else {
+            newValue = value.substring(0, length);
+        }
+        return new DatumString(newValue, PType.varchar(length));
```
Comment (alancai98) [inline partiql-spi/src/main/java/org/partiql/spi/value/Datum.java:666]: should be `character`?

```suggestion
        return new DatumString(newValue, PType.character(length));
```

**Review by alancai98 — APPROVED** [review]
pre-approving assuming that small typo is fixed

**Pull request merged**